app/models/miq_smart_proxy_worker/runner.rb

```diff
@@ -1,3 +1,73 @@
 class MiqSmartProxyWorker::Runner < MiqQueueWorkerBase::Runner
   self.delay_startup_for_vim_broker = true # NOTE: For smartproxy role
+
+  def do_before_work_loop
+    @tid = start_heartbeat_thread
+  end
+
+  def before_exit(message, _exit_code)
+    @exit_requested = true
+    #
+    # Stop the Heartbeat Thread
+    #
+    safe_log("#{message} Stopping Heartbeat Thread.")
+
+    #
+    # Wait for the Heartbeat Thread to stop
+    #
+    unless @tid.nil?
+      safe_log("#{message} Waiting for Heartbeat Thread to Stop.")
+      begin
+        @tid.join(worker_settings[:heartbeat_thread_shutdown_timeout])
+      rescue NoMethodError => join_err
+        safe_log(join_err)
+      end
+    end
+  end
+
+  def start_heartbeat_thread
+    @exit_requested    = false
+    @heartbeat_started = Concurrent::Event.new
+    _log.info("#{log_prefix} Starting Heartbeat Thread")
+
+    tid = Thread.new do
+      begin
+        heartbeat_thread
+      rescue => err
+        _log.error("#{log_prefix} Heartbeat Thread aborted because [#{err.message}]")
+        _log.log_backtrace(err)
+        Thread.exit
+      ensure
+        @heartbeat_started.set
+      end
+    end
+
+    @heartbeat_started.wait
+    _log.info("#{log_prefix} Started Heartbeat Thread")
+
+    tid
+  end
+
+  def heartbeat_thread
+    @heartbeat_started.set
+    until @exit_requested
+      heartbeat
+      sleep 30
+    end
+  end
+
+  def do_work
+    if @tid.nil? || !@tid.alive?
+      if !@tid.nil? && @tid.status.nil?
```

Comment:
`unless @tid.try(:status)`?

Reply:
Does this condition need to live inside the condition above?  It looks like it could be moved above the outer conditional and improve readability.